### PR TITLE
[Limit orders] store and widget

### PR DESCRIPTION
**Pull request by shoom3301**
# Summary

**Features:**
 - `jotai` store with persisting in the localStorage
 - set input/output currency into URL from user input
 - set widget state from URL
 - set input/output amounts into the store from user input

**Fixes and improvements:**
 - `useAllTokens()` - added useMemo() to avoid looped updates
 - fixed blurred background in the Limit orders page
 - moved `chainId` from query params to location paths, only for Limit orders page (`/#/swap?chain=goerli` -> `/#/5/limit-orders/DAI/USDC`)
 - persist input/output currencies into URL (now you can change currency symbol in URL and in will be automatically applied in the widget)

**Important:**
We DON'T update input/output currencies directly in the store, only through URL.
For example, if you want to change input currency, you should update only URL (`history.push(...)`) and it will be automatically set into the store.

  # To Test
Skip please, should be tested later

  # Background
`localStorage.setItem('enableLimitOrders', '1')` - run it in the browser console to enable Limit orders

### CODE REVIEW

File: src/cow-react/modules/limitOrders/containers/QuoteResolver/hooks/useQuoteRequestParams.ts [hunk 1]
```diff
@@ -0,0 +1,30 @@
+import { useLimitOrdersTradeState } from 'cow-react/modules/limitOrders/hooks/useLimitOrdersTradeState'
+import { useWeb3React } from '@web3-react/core'
+import { LegacyFeeQuoteParams as FeeQuoteParams } from 'api/gnosisProtocol/legacy/types'
+import { OrderKind } from '@cowprotocol/contracts'
+import { Token } from '@uniswap/sdk-core'
+import { parseUnits } from '@ethersproject/units'
+import { useMemo } from 'react'
+
+export function useQuoteRequestParams(): FeeQuoteParams | null {
```
Comment (shoom3301) [inline src/cow-react/modules/limitOrders/containers/QuoteResolver/hooks/useQuoteRequestParams.ts:9]: https://github.com/cowprotocol/cowswap/pull/1124#discussion_r983752396

> Limit orders won't need the fee. Just the price, but not the fee because it will be zero (once its implemented).
> 
> We might want to get the fee if they are rejecting our orders... (just temporary, once implemented they will start adding it)

@anxolin Currently, if we specify a low fee amount, then API response with the error:
```
description: "Order does not include sufficient fee"
errorType: "InsufficientFee"
```

**Comment by alfetopito** [comment]
Not high priority, but it's possible to select the same token on both sides:
Screen Shot 2022-09-30 at 14 06 16

File: src/cow-react/modules/limitOrders/containers/LimitOrdersWidget/index.tsx [hunk 2]
```diff
@@ -26,35 +51,51 @@ export function LimitOrdersWidget() {
...
+      updateLimitOrdersState({ recipient })
+    },
+    [updateLimitOrdersState]
+  )
+
+  const doTrade = useCallback(() => {
+    tradeContext && tradeFlow(tradeContext)
+  }, [tradeContext])
+
+  console.debug('RENDER LIMIT ORDERS WIDGET', { inputCurrencyInfo, outputCurrencyInfo })
```
Comment (nenadV91) [inline src/cow-react/modules/limitOrders/containers/LimitOrdersWidget/index.tsx:98]: If the plan is to remove these, maybe we should add `Todo:` comments

File: src/cow-react/modules/limitOrders/hooks/useSetupLimitOrdersState.ts [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+    const newState: Partial<LimitOrdersState> = chainIdWasChanged
+      ? getDefaultLimitOrdersState(tradeStateFromUrl.chainId)
+      : {
+          chainId: tradeStateFromUrl.chainId,
+          recipient: tradeStateFromUrl.recipient || state.recipient,
+          inputCurrencyId: tradeStateFromUrl.inputCurrencyId,
+          outputCurrencyId: tradeStateFromUrl.outputCurrencyId,
+        }
+
+    console.log('UPDATE LIMIT ORDERS STATE:', newState)
```
Comment (nenadV91) [inline src/cow-react/modules/limitOrders/hooks/useSetupLimitOrdersState.ts:38]: Same as another comment.

**Review by nenadV91 — APPROVED** [review]
Reviewed code

File: src/cow-react/modules/limitOrders/containers/QuoteResolver/hooks/useQuoteRequestParams.ts (hunk 1 above)
Comment (anxolin) [reply src/cow-react/modules/limitOrders/containers/QuoteResolver/hooks/useQuoteRequestParams.ts:9]: Yep, good to add for now. Also, I believe the backend will add the possibility to create limit orders (at least in goerli and for some whitelisted addresses)

File: src/cow-react/modules/limitOrders/containers/QuoteResolver/index.tsx [hunk 4]
```diff
@@ -0,0 +1,25 @@
+import { useSetAtom } from 'jotai'
+import { limitOrdersQuoteAtom } from '@cow/modules/limitOrders/state/limitOrdersQuoteAtom'
+import { useEffect } from 'react'
+import { useQuoteRequestParams } from '@cow/modules/limitOrders/containers/QuoteResolver/hooks/useQuoteRequestParams'
+import { getQuote } from '@cow/api/gnosisProtocol'
+
+export const QuoteResolver = () => {
```
Comment (anxolin) [inline src/cow-react/modules/limitOrders/containers/QuoteResolver/index.tsx:7]: Does it make sense to have the resolver inside an `updater` directory (`limitOrders/updater`) and maybe call it `QuoteUpdater` since components that return null and just do some side effects in the app, we call them updaters

**Review by anxolin — COMMENTED** [review]
There's a lot of hooks in this PR, and its already becoming hard to follow all the logic. 

i.e. `Check if there's two tokens with the same symbol`
https://github.com/cowprotocol/cowswap/blob/865ed198c8a871dd650e49c5cdc572446449d374/src/cow-react/modules/limitOrders/hooks/useAreThereTokensWithSameSymbol.ts#L5

One small thing we can do, is to add some docs to the hooks, with a short phrase of what they do. This way, we understand their scope. 

ubernit: For the hooks that don't return anything, maybe is nice to add `void` so its more explicit declaration. Feel free to ignore

File: src/cow-react/common/hooks/useAllTokensList.ts [hunk 5]
```diff
@@ -0,0 +1,9 @@
+import { Token } from '@uniswap/sdk-core'
+import { useMemo } from 'react'
+import { useAllTokens } from 'hooks/Tokens'
+
+export function useAllTokensList(): Token[] {
+  const allTokens = useAllTokens()
+
+  return useMemo(() => Object.values(allTokens), [allTokens])
```
Comment (alfetopito) [inline src/cow-react/common/hooks/useAllTokensList.ts:8]: Would the json stringify for the deps tactic help here?

File: src/cow-react/common/hooks/useTokenBySymbolOrAddress.ts [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+    if (!symbolOrAddress) {
+      return null
+    }
+
+    return (
+      tokens.find(
+        (item) => item.address.toLowerCase() === symbolOrAddress.toLowerCase() || item.symbol === symbolOrAddress
+      ) || null
+    )
+    // eslint-disable-next-line react-hooks/exhaustive-deps
```
Comment (alfetopito) [inline src/cow-react/common/hooks/useTokenBySymbolOrAddress.ts:18]: Why do you need this?
Maybe I'm not seeing something

File: src/cow-react/constants/routes.ts [hunk 7]
```diff
@@ -1,9 +1,11 @@
+export const LIMIT_ORDERS_PATH = 'limit-orders'
+
 // ENUM with routes
 export enum Routes {
   HOME = '/',
   SWAP = '/swap',
   SWAP_OUTPUT_CURRENCY = '/swap/:outputCurrency',
-  LIMIT_ORDER = '/limit-order',
+  LIMIT_ORDER = '/:chainId?/limit-orders/:inputCurrencyId?/:outputCurrencyId?',
```
Comment (alfetopito) [inline src/cow-react/constants/routes.ts:8]: I think I mentioned this in another PR: can you use the const from line 1 here?

File: src/cow-react/modules/limitOrders/containers/LimitOrdersMenuUpdater/index.tsx [hunk 8]
```diff
@@ -0,0 +1,7 @@
+import { useParameterizeLimitOrdersInMenu } from '../../hooks/useParameterizeLimitOrdersInMenu'
```
Comment (alfetopito) [inline src/cow-react/modules/limitOrders/containers/LimitOrdersMenuUpdater/index.tsx:1]: Path?

**Review by alfetopito — COMMENTED** [review]
Did not finish the review, will have to get back to it later

File: src/cow-react/common/hooks/useAllTokensList.ts (hunk 5 above)
Comment (shoom3301) [reply src/cow-react/common/hooks/useAllTokensList.ts:8]: It's not necessary, because we have it inside of `useAllTokens`

File: src/cow-react/common/hooks/useTokenBySymbolOrAddress.ts (hunk 6 above)
Comment (shoom3301) [reply src/cow-react/common/hooks/useTokenBySymbolOrAddress.ts:18]: Thanks! It's redundant

File: src/cow-react/constants/routes.ts (hunk 7 above)
Comment (shoom3301) [reply src/cow-react/constants/routes.ts:8]: It's not possible:
image

File: src/cow-react/modules/limitOrders/containers/LimitOrdersMenuUpdater/index.tsx (hunk 8 above)
Comment (shoom3301) [reply src/cow-react/modules/limitOrders/containers/LimitOrdersMenuUpdater/index.tsx:1]: I don't remember exactly but seems like we discussed with @anxolin that we should use relative paths in the module scope. I don't have a strong opinion about it.

File: src/cow-react/modules/limitOrders/containers/QuoteResolver/index.tsx (hunk 4 above)
Comment (shoom3301) [reply src/cow-react/modules/limitOrders/containers/QuoteResolver/index.tsx:7]: Thank you!
Yes, this is updater

File: src/cow-react/modules/limitOrders/pure/Chart/index.cosmos.tsx [hunk 9]
```diff
@@ -0,0 +1,3 @@
+import { ChartPure } from './index'
```
[the remark this answers is not in the document]
Comment (shoom3301) [reply src/cow-react/modules/limitOrders/pure/Chart/index.cosmos.tsx:1]: I guess @nenadV91 did it to distinguish two Chart components: from containers and from pure.
I suggest to use Widget suffix for containers:
`pure/Chart`
`containers/ChartWidget`

File: src/cow-react/modules/limitOrders/hooks/useParameterizeLimitOrdersInMenu.ts [hunk 10]
```diff
@@ -0,0 +1,34 @@
...
+
+export function useParameterizeLimitOrdersInMenu() {
+  const { chainId } = useWeb3React()
+  const { inputCurrencyId, outputCurrencyId } = useAtomValue(limitOrdersAtom)
+  const overrideMainMenu = useSetAtom(mainMenuUrlOverridesAtom)
+
+  useEffect(() => {
+    const route = parameterizeLimitOrdersRoute(chainId, inputCurrencyId, outputCurrencyId)
+
+    overrideMainMenu({ [MainMenuItemId.LIMIT_ORDERS]: route })
```
[the remark this answers is not in the document]
Comment (shoom3301) [reply src/cow-react/modules/limitOrders/hooks/useParameterizeLimitOrdersInMenu.ts:18]: And it would be weird to make `Header` dependent on `Limit-orders`. So, using `mainMenuUrlOverridesAtom` we avoid of `Dependency inversion` principle breach.

File: src/cow-react/modules/limitOrders/hooks/useLimitOrdersTradeState.ts [hunk 11]
```diff
@@ -0,0 +1,39 @@
...
+  readonly deadline: number | null
+}
+
+export function useLimitOrdersTradeState(): LimitOrdersTradeState {
+  const state = useAtomValue(limitOrdersAtom)
+
+  const recipient = state.recipient
+  const deadline = state.deadline
+  const inputCurrency = useTokenBySymbolOrAddress(state.inputCurrencyId) || null
+  const outputCurrency = useTokenBySymbolOrAddress(state.outputCurrencyId) || null
```
Comment (alfetopito) [inline src/cow-react/modules/limitOrders/hooks/useLimitOrdersTradeState.ts:23]: `useTokenBySymbolOrAddress` already returns Token | null, so you don't need `|| null` here, do you?

**Review by alfetopito — APPROVED** [review]
Some minor comments, behaviour seems fine except for when the token doesn't exist.
Still that should be addressed in another PR as this one is already quite big

File: src/cow-react/modules/limitOrders/hooks/useResetStateWithSymbolDuplication.ts [hunk 12]
```diff
@@ -0,0 +1,44 @@
...
+) => t`There is more than one token in the list of tokens with the symbol: ${doubledSymbol}.
+Please select the token you need from the UI or use the address of the token instead of the symbol`
+
+/**
+ * Case: when user opened a link with a token symbol and there are more than one token with the same symbol
+ * Example: /limit-orders/UST/WETH
+ * https://etherscan.io/token/0xa47c8bf37f92abed4a126bda807a7b7498661acd - Symbol: UST
+ * https://etherscan.io/token/0xa693b19d2931d498c5b318df961919bb4aee87a5 - Symbol: UST
+ * In this case we just reset state to default, because:
+ * if user selected a token with symbol duplication from cowswap Dapp, then Dapp puts a token address in the URL
```
[the remark this answers is not in the document]
Comment (shoom3301) [reply src/cow-react/modules/limitOrders/hooks/useResetStateWithSymbolDuplication.ts:21]: Yes, you are right, I will add a task for it

File: src/cow-react/modules/limitOrders/hooks/useTradeFlowContext.ts [hunk 13]
```diff
@@ -0,0 +1,73 @@
...
+    !state.deadline ||
+    !provider ||
+    !settlementContract ||
+    !appData
+  ) {
+    return null
+  }
+
+  const isGnosisSafeWallet = !!gnosisSafeInfo
+  const validTo = Math.round(Date.now() / 1000 + 60 * state.deadline)
```
[the remark this answers is not in the document]
Comment (shoom3301) [reply src/cow-react/modules/limitOrders/hooks/useTradeFlowContext.ts:43]: Sorry, I don't understand

File: src/cow-react/modules/limitOrders/updaters/QuoteUpdater/index.tsx [hunk 14]
```diff
@@ -0,0 +1,25 @@
+import { useSetAtom } from 'jotai'
+import { limitOrdersQuoteAtom } from '@cow/modules/limitOrders/state/limitOrdersQuoteAtom'
+import { useEffect } from 'react'
+import { useQuoteRequestParams } from './hooks/useQuoteRequestParams'
+import { getQuote } from '@cow/api/gnosisProtocol'
+
+export const QuoteUpdater = () => {
```
[the remark this answers is not in the document]
Comment (shoom3301) [reply src/cow-react/modules/limitOrders/updaters/QuoteUpdater/index.tsx:7]: Thanks! I also prefer using functions instead of const :)

File: src/cow-react/modules/limitOrders/hooks/useTradeFlowContext.ts (hunk 13 above)
[the remark this answers is not in the document]
Comment (alfetopito) [reply src/cow-react/modules/limitOrders/hooks/useTradeFlowContext.ts:43]: I mean, add a TODO so when we start to actually implement the order duration in the UI, we know this is a place where it's hard coded.
